### ISSUE TITLE
Passing an updated fetcher method has no effect

### ISSUE THREAD

**Issue by PMudra** [vercel/swr#1131]
# Bug report

## Description / Observed Behavior

Passing an updated fetcher method to `useSWR` has no effect. The initial version of the fetcher will be used.

I am passing a fetcher function to `useSWR` that is updated by applying a new access token.

```js
const [accessToken, setAccessToken] = useState("valid-0");
const fetcher = () => {
  console.log("fetch with token:", accessToken);
};
const swrWithOutdatedFetcher = useSWR("key", fetcher);
```

When I call `setAccessToken("valid-1")` at some point (and revalidate), the output will still be `"fetch with token: valid-0"`.

## Expected Behavior

I was expecting swr to use the updated fetcher method. I believe this is the expected way when working with hooks in react.

In my example I'd expect `"fetch with token: valid-1"`.

The workaround for now is to use `useRef` in combination with `useEffect` and then call `fetcherRef.currect()`.

```js
const fetcherRef = useRef(fetcher);
useEffect(() => {
  fetcherRef.current = fetcher;
});
const swrWithUpdatedFetcher = useSWR("key", () => fetcherRef.current());
```

If this behavior is intended, some words about this is the docs would be helpful and appreciated.

## Repro Steps / Code Example

A simple example to reproduce and also the workaround using `useRef`:

https://codesandbox.io/s/autumn-butterfly-c4xe2?file=/src/App.tsx

## Additional Context

SWR version 0.5.5

Related: #515 #927 https://github.com/vercel/swr/blob/3d380e21fed1f41b7df7816c55216945bf80eae9/src/use-swr.ts#L479-L486

**Comment by shuding** [comment]
Changing `fetcher` during render might cause unexpected behavior, so it will be ignored and won't trigger a refetch. This is simply because people can pass e.g. anonymous function as the fetcher that will be re-declared on every render. In SWR, `key` is always the identifier of the resource.

**Comment by PMudra** [comment]
I agree that a changed `fetcher` should not trigger a refetch. What I expected is that the next refetch uses the latest `fetcher` I have passed to the hook.

Passing the token as a `key` to the hook looks promising at first thought. But that would lead to an immediate refetch of all resources which I do not want. The data hasn't changed. Just the way to get it (the next time I need it).

**Comment by Gyllsdorff** [comment]
> Changing fetcher during render might cause unexpected behavior, so it will be ignored and won't trigger a refetch.

> I agree that a changed fetcher should not trigger a refetch. What I expected is that the next refetch uses the latest fetcher I have passed to the hook.

➕1 on that. FWIW; I know several developers that didn't understand that the fetcher instance that was *first* used as the [fetcher for that set of keys (and component?)](https://github.com/vercel/swr/blob/0.5.6/src/use-swr.ts#L530-L534) will be used for all/most subsequent request (for that set of keys).

Is it possible to add some information about this to the [Data fetching](https://swr.vercel.app/docs/data-fetching) page? It feels like 99% of the time when we pass a fetcher directly to `useSWR` we do it the wrong way, it is likely better to configure the fetch using the global configuration.

**Comment by ivan-kleshnin** [comment]
> In SWR, key is always the identifier of the resource.

And how to refetch certain key with new params? I tried `cache.delete(key)`, `mutate(key)`, tried to change fetching function like @PMudra – nothing helps 😞  

When that may be necessary? Fetching indexes with different filters, sorting, pagination. By using different keys tons of index data is accumulated in memory and its hard to clear it up in a library-level code (where you don't know those fetching details).
In other words I would like to have all requests to the same URL e.g `/api/todos` to be tied to a single known key (`/api/todos`) and new data to replace old data in memory, not to add incrementally.

**Comment by steida** [comment]
I wish SWR was a resilient Hook. useEffect with ref is an ugly hack of broken API.

https://overreacted.io/writing-resilient-components/

**Comment by ivan-kleshnin** [comment]
> It feels like 99% of the time when we pass a fetcher directly to useSWR we do it the wrong way, it is likely better to configure the fetch using the global configuration.

This is not true. We use local fetchers most of the time:

```
export const useFetchJobTalents = (
  params : FetchJobTalentsParams, query : FetchJobTalentsQuery, options
) : SWRResponse<FetchJobTalentsResult> => {
  options = ... // omitted for brevity

  const cacheKey = [fetchJobTalents.url, params.jobId, stringify(query)]
  useSWR_GarbageCollection(cacheKey)

  return useSWR(cacheKey, () => {
    return fetchJobTalents(params, query) // custom fetcher
  }, options)
}
```

**Comment by zlace0x** [comment]
Also faced this as an issue, this is not a given reading the docs.

**Comment by 0xMAJIMAK** [comment]
Have also faced this issue, given the nature of hooks it would be assumed that if new instances of the fetcher would be used when the key changes.

**Comment by shuding** [comment]
Would love to accept a PR to improve this! 👍
